p2p/src/network.rs

```diff
@@ -2,516 +2,451 @@
-    K: KeyExchangeScheme + Send + 'static,
-    E: Encryptor + Send + 'static,
-{
-    type Result = ();
-
-    async fn handle(&mut self, msg: ConnectPeer) {
-        if self.peers.contains_key(&msg.peer.public_key) {
-            debug!(peer = %msg.peer, "Peer already connected");
+    fn connect_peer(&mut self, ConnectPeer { peer_id: peer }: ConnectPeer) {
+        if self.peers.contains_key(&peer.public_key) {
+            iroha_logger::debug!(peer = %peer, "Peer already connected");
             return;
         }
 
-        debug!(
-            listen_addr = %self.listen_addr, peer.id.address = %msg.peer.address,
+        iroha_logger::debug!(
+            listen_addr = %self.listen_addr, peer.id.address = %peer.address,
             "Creating new peer actor",
         );
-        self.untrusted_peers.remove(&ip(&msg.peer.address));
-        let peer_to_key_exchange = match Peer::new_to(
-            PeerId::new(&msg.peer.address, &self.public_key),
-            self.broker.clone(),
-        )
-        .await
-        {
-            Ok(peer) => peer,
-            Err(error) => {
-                warn!(%error, "Unable to create peer");
-                return;
-            }
-        };
-        #[allow(clippy::expect_used)]
-        let conn_id = peer_to_key_exchange
-            .connection_id()
-            .expect("has connection by construction.");
-        let addr = peer_to_key_exchange.start().await;
-        debug!(%conn_id, ?addr, "Inserting into new_peers");
-        self.new_peers.insert(conn_id, addr.clone());
-        addr.do_send(Start).await;
+        self.untrusted_peers.remove(&ip(&peer.address));
+
+        let conn_id = rand::random();
+        let (connected_sender, connected_receiver) = oneshot::channel();
+        let (terminated_sender, terminated_receiver) = oneshot::channel();
+        connecting::<T, K, E>(
+            PeerId::new(&peer.address, &self.public_key),
+            conn_id,
+            connected_sender,
+            terminated_sender,
+        );
+
+        self.connected_receivers.push(connected_receiver);
+        self.terminated_receivers.push(terminated_receiver);
     }
-}
 
-#[async_trait::async_trait]
-impl<T, K, E> Handler<DisconnectPeer> for NetworkBase<T, K, E>
-where
-    T: Debug + Encode + Decode + BrokerMessage + Send + Sync + Clone + 'static,
-    K: KeyExchangeScheme + Send + 'static,
-    E: Encryptor + Send + 'static,
-{
-    type Result = ();
-
-    // For some reason basic let-else here violets the borrow checker rules
-    #[allow(clippy::manual_let_else)]
-    async fn handle(&mut self, DisconnectPeer(public_key): DisconnectPeer) {
+    fn disconnect_peer(&mut self, DisconnectPeer(public_key): DisconnectPeer) {
         let peer = match self.peers.remove(&public_key) {
             Some(peer) => peer,
-            _ => return error!(?public_key, "Not found peer to disconnect"),
+            _ => return iroha_logger::error!(?public_key, "Not found peer to disconnect"),
```

Comment:
Both `warn` and `error` indicate problems. For each `error` there must be a clear set of instructions for the peer admin to follow. What is the corresponding set of instructions for a peer not being found for disconnection. There's not a lot you can do, other than `warn` the peer admin that something's not right.

Reply:
so you see `error` as actionable. This makes sense to me unless it's due to a bug in the system. There the line is not so clear. For me `warn` is saying something like: this problem happened but don't you worry too much because there is some automatic recovery in place that will take care of it for you or the problem that was witnessed is not detrimental to the system operation. In the case of a malfunctioning system, I think we should use `error` in the logs or just bring the node down. Arguably, critical bug should never happen (at least it's unlikely to happen) in our code

I agree this looks like it should be a `warn`. It's not a critical issue, nor is it actionable